infra/src/index.ts

```diff
@@ -17,12 +35,88 @@ export default class S3Webhosting extends MiraStack {
     const siteBucket = new AutoDeleteBucket(this, 'SiteBucket', bucketProps)
     this.addOutput('WebsiteURL', siteBucket.bucketWebsiteUrl)
 
-    const webAppPath = path.join(__dirname, '..', '..', 'web-app')
+    const config = MiraConfig.getEnvironment() as AccountWithDomainSettings
+
+    let distributionDomainName
+    if (config.withDomain) {
+      if (!config.webAppUrl) {
+        throw new Error('"webAppUrl" config is required when "withDomain" is true')
+      }
+      distributionDomainName = this.getFullDeployment(siteBucket, config.webAppUrl)
+    } else {
+      distributionDomainName = this.getMinimalDeployment(siteBucket)
+    }
+    this.addOutput('distributionDomainName', distributionDomainName)
+  }
+
+  private getMinimalDeployment (siteBucket: AutoDeleteBucket): string {
+    const distribution = this.getCloudFrontDistribution(siteBucket)
+    
+    this.createBucketDeployment(siteBucket, distribution)
+
+    return distribution.distributionDomainName
+  }
+
+  private getFullDeployment (siteBucket: AutoDeleteBucket, webAppUrl: string): string {
+    const certificate = new CustomCertificate(this, {
+      domain: webAppUrl
+    })
+
+    const distribution = this.getCloudFrontDistribution(siteBucket, {
+      certificateArn: certificate.certificateArn,
+      webAppUrl
+    })
+
+    new CustomDomain(this, {
+      source: webAppUrl,
+      target: distribution.distributionDomainName
+    })
+
+    return distribution.distributionDomainName
+  }
+
+  createBucketDeployment(siteBucket: AutoDeleteBucket, distribution: CloudFrontWebDistribution) {
     new BucketDeployment(this, 'Deployment', {
       destinationBucket: siteBucket,
+      distribution,
+      distributionPaths: ['/*'],
       sources: [
-        S3DeploymentSource.asset(webAppPath)
+        S3DeploymentSource.asset(path.join(__dirname, '..', '..', 'web-app'))
       ]
     })
   }
+
+  private getCloudFrontDistribution(siteBucket: any, certificateConfig?: CertificateConfigProps) {
```

Comment:
With an item like this seems like it may be able to live outside this class as it is specific to Cloudfront and could be reusable?  May be out of scope for this PR.

Reply:
Yes, probably worth moving outside although it is a very small example and can be easier to track everything in one file.